### ISSUE TITLE
Add an option to shutdown when connections reach zero

### ISSUE THREAD

**Issue by huggsboson** [envoyproxy/envoy#3837]
*Description*:
For graceful shutdown scenarios described in #1990  we largely have what we need, but it would be extra helpful to allow envoy to exit once ingress connections reach zero.

Basically if we have a service that wants to:
1. Use envoy for ingress and egress
2. Has a potentially long and highly variable drain time (e.g. 5 min to 4 hours)

It would be nice if we could have a sigterm:
1. unbind from ingress ports and begin draining the connections.
2. leave egress open for any inflight requests 
Note: #1 and #2 are already supported
3. Once all of the ingress (or all!) connections have closed then have envoy exit.

*Relevant Links*:
https://github.com/envoyproxy/envoy/issues/1990

**Comment by huggsboson** [comment]
I don't think this is stale, but I obviously haven't implemented it.  Do you generally close things that are enhancements, but haven't been worked on?

**Comment by markusthoemmes** [comment]
Yes! I feel like this is an important feature especially in scenarios like Kubernetes where containers are generally killed via SIGTERM. If we'd handle that gracefully as described, that'd be super helpful. Anything I can help with here?

**Comment by markusthoemmes** [comment]
@huggsboson Is the restarter script maybe an option for you? https://github.com/envoyproxy/envoy/blob/master/restarter/hot-restarter.py

**Comment by huggsboson** [comment]
I can't quite tell the name of the script seems to point at restarting as the goal, but can it assist with shutdowns as well?

**Comment by huggsboson** [comment]
Reading over the script, I don't think it will. What I need is the main executable itself to allow SIGTERM to trigger a very long drain (4 hrs) but to let it exit if all existing connections close sooner.

**Comment by markusthoemmes** [comment]
I agree, wrote the comment at a point in time where I didn‘t have as much context. I think we should implement graceful shutdown when SIGTERM is sent.

**Comment by wrowe** [comment]
SIGTERM is typically the signal sent by the OS during a system shutdown, so it is unwise to use this anything slower than such rapid (<120 second) shutdown windows. Both httpd and ngnix use the WINCH signal for this slower, graceful shutdown purpose.

**Comment by jmarantz** [comment]
See also https://github.com/istio/istio/pull/35059

**Comment by huggsboson** [comment]
The specific ask here is to close the ingress port to new connections on SIGTERM, but leave any egress ports open for reqeusts that are inflight.  Then exit after the ingress port's open connections/requests reaches zero.  I can't quite tell if that PR is supporting this or not.

**Comment by huggsboson** [comment]
Actually this is super helpful and we're adopting it internally to address the issue I mentioned originally.  Going to close this now.

**Comment by jmarantz** [comment]
You opened this issue in Envoy, and I think it's still open in Envoy :)

There's an Istio-specific solution to this, and we can either leave this open or duplicate it because for our use-case we'd like it natively in Envoy.

**Comment by kyessenov** [comment]
Speaking as Istio dev, I'd prefer if this was solved in Envoy :) The termination logic is as follows:
1. SIGTERM puts "INBOUND" listeners into draining mode and issuing GOAWAY. Connections are accepted. Clients take time to re-balanced.
2. Fixed delay, to prevent a race between application receiving SIGTERM and opening a connection.
3. Agent watches the stat for "OUTBOUND" connections, when one of the following happens
3a. k8s deadline for graceful termination expires, SIGKILL.
3b. Agent counts the stat to 0, stops Envoy.

Note that `drain-duration` flag in Envoy is counter-productive, since Envoy generally doesn't know how much is allotted by the platform.

**Comment by jmarantz** [comment]
Thanks Kuat. Out of curiosity why is it better to accept connections and issue GOAWAY as opposed to not accepting connections? Also is that assuming H2? What about H1 and TCP?

**Comment by huggsboson** [comment]
I've spent some time thinking about watching the outbound side as a trigger condition for closing and I don't think it works fully.  You could have a request inflight and close your outbound connections and then open a new one a half second later as part of the same inbound request.  I think you need to watch the inbound side for activity to go to zero.

**Comment by kyessenov** [comment]
@jmarantz I believe that's the standard way to implement lame duck mode, for "drain for termination". Sometimes, there's an intermediate mode when "termination" mode kicks in that also cancels in-flight requests and closes sockets. With both directions of traffic, it's rather complicated to introduce another state.

**Comment by jmarantz** [comment]
For L7 proxies I was thinking of tracking inbound request-count not connections. I think inbound might be easier because I would usually expect fewer listeners than upstream connections. Though (at least naively) the number of inbound and outbound requests should be the same, right?

**Comment by kyessenov** [comment]
@jmarantz Yeah, that makes sense for a gateway. It's important to accept for some time after drain initiated since there's a config propagation delay for the gateway to be fully removed from the pools.

What I referred by the OUTBOUND is not upstream connection, it's downstream but coming from the sidecar backend. These are things like backend dumping logs right before exiting.

I think there's really two distinct phases that we need:
* drain for termination: continuing as normal but send GOAWAY
* termination: stop accepting, cancel in-flight, wait for in-flight (non health check) requests to go to 0, exit.

I think the latter phase is what's missing in Envoy.

**Comment by jmarantz** [comment]
Makes sense I think. RE upstream/downstream -- is the point that as a sidecar there's only one backend but perhaps multiple listeners, so checking backend connections is cheaper?

RE "cancel in-flight" -- what do you mean by that? I thought we'd give in-flight requests some time-limit to finish before we cancel them.
